srtcore/core.h

```diff
@@ -144,6 +156,379 @@ class CCryptoControl;
+
+    int send(const char* buf, int len, SRT_MSGCTRL& w_mc);
+    int recv(char* buf, int len, SRT_MSGCTRL& w_mc);
+
+    void close();
+
+    void setOpt(SRT_SOCKOPT optname, const void* optval, int optlen);
+    void getOpt(SRT_SOCKOPT optName, void* optval, int& w_optlen);
+    void deriveSettings(CUDT* source);
+
+    SRT_SOCKSTATUS getStatus();
+
+    bool getMasterData(SRTSOCKET slave, SRTSOCKET& w_mpeer, time_point& w_st);
+
+    bool isGroupReceiver()
+    {
+        // XXX add here also other group types, which
+        // predict group receiving.
+        return m_type == SRT_GTYPE_BROADCAST;
+    }
+
+    srt::sync::Mutex* exp_groupLock() { return &m_GroupLock; }
+    void addEPoll(int eid);
+    void removeEPoll(int eid);
+    void updateReadState(SRTSOCKET sock, int32_t sequence);
+    void updateWriteState();
+
+    /// Update the in-group array of packet providers per sequence number.
+    /// Also basing on the information already provided by possibly other sockets,
+    /// report the real status of packet loss, including packets maybe lost
+    /// by the caller provider, but already received from elsewhere. Note that
+    /// these packets are not ready for extraction until ACK-ed.
+    ///
+    /// @param exp_sequence The previously received sequence at this socket
+    /// @param sequence The sequence of this packet
+    /// @param provider The core of the socket for which the packet was dispatched
+    /// @param time TSBPD time of this packet
+    /// @return The bitmap that marks by 'false' packets lost since next to exp_sequence
+    std::vector<bool> providePacket(int32_t exp_sequence, int32_t sequence, CUDT *provider, uint64_t time);
+
+    /// This is called from the ACK action by particular socket, which
+    /// actually signs off the packet for extraction.
+    ///
+    /// @param core The socket core for which the ACK was sent
+    /// @param ack The past-the-last-received ACK sequence number
+    void readyPackets(CUDT* core, int32_t ack);
+
+    void syncWithSocket(const CUDT& core);
+    int getGroupData(SRT_SOCKGROUPDATA *pdata, size_t *psize);
+
+    /// Predicted to be called from the reading function to fill
+    /// the group data array as requested.
+    void fillGroupData(
+            SRT_MSGCTRL& w_out, //< MSGCTRL to be written
+            const SRT_MSGCTRL& in, //< MSGCTRL read from the data-providing socket
+            SRT_SOCKGROUPDATA* out_grpdata, //< grpdata as passed in MSGCTRL
+            size_t out_grpdata_size  //< grpdata_size as passed in MSGCTRL
+            );
+
+#if ENABLE_HEAVY_LOGGING
+    void debugGroup();
+#else
+    void debugGroup() {}
+#endif
+private:
+    // Check if there's at least one connected socket.
+    // If so, grab the status of all member sockets.
+    void getGroupCount(size_t& w_size, bool& w_still_alive);
+    void getMemberStatus(std::vector<SRT_SOCKGROUPDATA>& w_gd, SRTSOCKET wasread, int result, bool again);
```

Comment:
const?

Reply:
There's a problem here. This isn't any part of the external function anyway (it's private), while in order to make it const I'd have to introduce lots of other changes in this class. For the beginning:
 - make the mutex mutable
 - introduce the `cbegin()` and `cend()` functions for the container
and I haven't even checked everything. Might be that there is some function doing a passthrough, and then it will cost it much more gymnastics.